report/4.Solution/SemanticsDescription.tex

```diff
@@ -0,0 +1,762 @@
+\end{center}
+
+\texttt{[$Dcl_VAssFnc_{BSS}$]}\\
+\begin{center}
+	\begin{math}
+		\cfrac
+			{\langle D_V,\ env_V[x \mapsto l][next \mapsto l'],\ sto[l \mapsto v] \rangle \rightarrow_{D_V} (env_V',\ sto')}
+			{\langle T\ x = f;\ S,\ env_V,\ sto \rangle \rightarrow_{D_V} (env_V',\ sto')}		
+	\end{math}
+	
+	\texttt{ where } $env_V,\ sto \vdash f \rightarrow_{call_{exp}} v$
+	\texttt{ and } $l = env_V\ next$
+	\texttt{ and } $l' = new\ l$
+\end{center}
+
+\textbf{[$Dcl_VNoAss_{BSS}$]}\\
+\begin{center}
+	\begin{math}
+	\cfrac
+		{\langle D_v,\ env_v[x \mapsto l][next \mapsto l'],\ sto \rangle \rightarrow_{D_V} (env_v',\ sto)}
+		{\langle T\ x;\ S,\ env_v,\ sto \rangle \rightarrow_{D_V} (env_v',\ sto)}
+	\end{math}
+	
+	\texttt{ where } $l = env_v\ next$
+	\texttt{ and } $l' = new\ l$
+\end{center}
+
+\textbf{[$Dcl_VEmpty_{BSS}$]}\\
+\begin{center}
+	\begin{math}
+	\langle \epsilon,\ env_v,\ sto \rangle \rightarrow (env_v,\ sto)
+	\end{math}
+\end{center}
+
+\subsection{Big-Step Semantics for Procedure and Function Declarations}
+Function declarations $Dcl_P$ can change the procedure environment $Env_P$ by declaring new functions in $Fnames$.
+
+The transition-system for $Dcl_P$ is ($Fnames \rightharpoonup Lambda\ X\ Var\ X\ Exp\ X\ Env_V\ X\ Env_P$).
+This means that the transitions must be on the form.
+
+\begin{center}
+	$\langle D_P, env_V, env_P \rangle \rightarrow_{D_P} (env_P)$
+\end{center}
+
+Procedures are functions with no return value and therefore are statements.
+Any function returning a value is an expression.
+For simplicity, all functions will be denoted f in the function declaration transitions.
+
+Declarations of functions must remember the expression directly following the return keyword, this is accomplished by binding the function in the procedure environment with the value of this expression.
+This is not the case for procedures, since they do not return a value.
+The rules for procedures and functions are the same, except for the value $v$ being $none$ in procedure rules.
+
+The transitions are shown below.
+
+\texttt{[$Dcl_PAss_{BSS}$]}\\
+\begin{center}
+	\begin{math}
+		\cfrac
+			{env_V \vdash \langle L, env_P[f \vdash (L, x\sim, v, env_V, env_P)] \rangle \rightarrow_{D_P} env_P'}
+			{env_v \vdash \langle T\ f\ =\ ([T\ x]\sim):\ T\ L;\ S,env_P \rangle \rightarrow_{D_P} env_P'}
+	\end{math}
+	
+	\texttt{ where } $env_V, sto \vdash L \rightarrow_L v$
+\end{center}
+
+\texttt{[$Dcl_PNoAss_{BSS}$]}\\
+\begin{center}
+	\begin{math}
+	\cfrac
+	{env_V \vdash \langle L, env_P[f \vdash (env_V, env_P)] \rangle \rightarrow_{D_P} env_P'}
+	{env_v \vdash \langle T\ f;\ S,env_P \rangle \rightarrow_{D_P} env_P'}
+	\end{math}
+\end{center}
+
+\texttt{[$PreDcl_Ass_{BSS}$]}\\
+\begin{center}
+	\begin{math}
+	\cfrac
+	{env_V \vdash \langle L, env_P[f \vdash (L, x\sim, v, env_V, env_P)] \rangle \rightarrow_{D_P} env_P'}
+	{env_v \vdash \langle f\ =\ ([T\ x]\sim):\ T\ L;\ S,env_P \rangle \rightarrow_{D_P} env_P'}
```

Comment:
Samme kommentar som ovenstående
Måske nævn, at reglen er for assignment af en funktion, der allerede er erklæret.